contracts/6/forwarder/ERC20Forwarder.sol

```diff
@@ -192,6 +193,72 @@ import "./BiconomyForwarder.sol";
             emit FeeCharged(req.from,charge,req.token);
     }
 
+    /**
+     * @dev
+     * - calls permit method on the underlying ERC20 token contract (DAI permit type) with given permit options
+     * - Keeps track of gas consumed
+     * - Calls BiconomyForwarder.executeEIP712 method using arguments given
+     * - Calls _transferHandler, supplying the gas usage of the executeEIP712 call
+     */
+    /**
+     * @param req : the request being forwarded
+     * @param domainSeparator : the domain separator presented to the user when signing
+     * @param sig : the signature generated by the user's wallet
+     * @param permitOptions : the permit request options for executing permit 
+     * @return success : false if call fails. true otherwise
+     * @return ret : any return data from the call
+     */
+    function permitAndExecuteEIP712(
+        ERC20ForwardRequest memory req,
+        bytes32 domainSeparator,
+        bytes calldata sig,
+        PermitRequest memory permitOptions
+        )
+        external payable
+        returns (bool success, bytes memory ret){
+            uint256 initialGas = gasleft();
+            //DAI permit
+            IERC20Permit(req.token).permit(permitOptions.holder, permitOptions.spender, permitOptions.nonce, permitOptions.expiry, permitOptions.allowed, permitOptions.v, permitOptions.r, permitOptions.s);
+            (success,ret) = BiconomyForwarder(forwarder).executeEIP712(req,domainSeparator,sig);
+            uint256 postGas = gasleft();
+            uint256 transferHandlerGas = transferHandlerGas[req.token];
+            uint256 charge = _transferHandler(req,initialGas.add(baseGas).add(transferHandlerGas).sub(postGas));
+            emit FeeCharged(req.from,charge,req.token);
+    }
+
+    /**
+     * @dev
+     * - calls permit method on the underlying ERC20 token contract (which supports EIP2612 permit) with given permit options
+     * - Keeps track of gas consumed
+     * - Calls BiconomyForwarder.executeEIP712 method using arguments given
+     * - Calls _transferHandler, supplying the gas usage of the executeEIP712 call
+     */
+    /**
+     * @param req : the request being forwarded
+     * @param domainSeparator : the domain separator presented to the user when signing
+     * @param sig : the signature generated by the user's wallet
+     * @param permitOptions : the permit request options for executing permit 
+     * @return success : false if call fails. true otherwise
+     * @return ret : any return data from the call
+     */
+    function permitEIP2612AndExecuteEIP712(
+        ERC20ForwardRequest memory req,
+        bytes32 domainSeparator,
+        bytes calldata sig,
+        PermitRequest memory permitOptions
```

Comment:
Does EIP2612 has same structure as dai permit?

Reply:
They have common fields but different in fields value and allowed.
So I have created common struct 
If a field is not used pass mock value in struct.
But I get your point, should I create two different structs?